contracts/eosio.system/src/finalizer_key.cpp

```diff
@@ -0,0 +1,158 @@
+#include <eosio.system/eosio.system.hpp>
+
+#include <eosio/eosio.hpp>
+#include <eosio/privileged.hpp>
+#include <eosio/crypto_bls_ext.hpp>
+
+namespace eosiosystem {
+
+   bool system_contract::is_savanna_consensus() {
+      return !_gstate5.last_finalizers.empty();
+   }
+
+   bool system_contract::is_registered_producer( const name& producer ) {
+      // Returns true if `producer` is a registered producer
+   }
+
+   bool system_contract::is_active_producer( const name& producer ) {
+      // Returns true if `producer` is an active producer (top 21)
+   }
+
+   bool set_next_finalizers_set(const std::vector<name>& producers ) {
+      // Build finalizer set
+      std::vector<finalizer_authority> finalizers;
+      for ( const auto p&: producers ) {
+         auto it = _finalizer_keys.find( p.value );
+         if( it != _finalizer_keys.end() ) { // p has registered a finalizer key
+            finalizers.emplace_back(
+               finalizer_authority{
+                  .description = "",
+                  .weight      = 1,
+                  .public_key  = it->active_key
+               }
+            );
+         }
+      }
+
+      // Abort if less than 15 top producers have registered a finalizer key
+      if( finalizers.size < 15 ) {
+         return false;
+      }
+
+      // Establish finalizer policy and call set_finalizers() host function
+      finalizer_policy fin_policy {
+         .generation = 0;
```

Comment:
Why are we required to provide a generation number?

I don't think we should require that as part of the `set_finalizers` host function interface. It puts extra burden on the caller to make sure they get the number right? What if it is wrong? Will the host function abort?

It makes it too complicated to guarantee we get that number right, especially if you consider that finalizer policies proposed during the transition from legacy to IF will need to be dropped. And we have no way of knowing that as a caller.

Reply:
This was my mistake. Generation number is not required.